compiler/src/iree/compiler/Codegen/Common/VectorizePackUnPackOps.cpp

```diff
@@ -0,0 +1,152 @@
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/Utils/IndexingUtils.h"
+#include "mlir/IR/AffineMap.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-codegen-vectorize-pack-unpack-ops"
+
+namespace mlir {
+namespace iree_compiler {
+namespace {
+
+/// A simple pattern rewriter that implements no special logic.
+class SimpleRewriter : public PatternRewriter {
+ public:
+  SimpleRewriter(MLIRContext *context) : PatternRewriter(context) {}
+};
+
+struct VectorizePackUnPackOpsPass
+    : public VectorizePackUnPackOpsBase<VectorizePackUnPackOpsPass> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<linalg::LinalgDialect, func::FuncDialect,
+                    arith::ArithDialect, scf::SCFDialect, tensor::TensorDialect,
+                    vector::VectorDialect>();
+  }
+
+  void runOnOperation() override {
+    MLIRContext *ctx = &getContext();
+    auto funcOp = getOperation();
+
+    // Apply tiling to make outer dims be all 1s.
+    {
+      SimpleRewriter rewriter(ctx);
+      auto packOptions = scf::SCFTileAndFuseOptions().setTilingOptions(
+          scf::SCFTilingOptions().setTileSizeComputationFunction(
+              [](OpBuilder &builder, Operation *op) -> SmallVector<Value> {
+                auto packOp = cast<tensor::PackOp>(op);
+
+                // Do nothing if any of inner tile sizes is dynamic.
+                if (llvm::any_of(packOp.getMixedTiles(), [](OpFoldResult tile) {
+                      return tile.is<Value>();
+                    })) {
+                  return {};
+                }
+
+                int inputRank = packOp.getSourceRank();
+                SmallVector<Value> tileSizes(
+                    inputRank,
+                    builder.create<arith::ConstantIndexOp>(op->getLoc(), 1));
+                return tileSizes;
+              }));
+      funcOp->walk([&](tensor::PackOp op) {
+        FailureOr<scf::SCFTileAndFuseResult> tileAndFuseResult =
+            scf::tileConsumerAndFuseProducerGreedilyUsingSCFForOp(
+                rewriter, cast<TilingInterface>(op.getOperation()),
+                packOptions);
+        if (failed(tileAndFuseResult)) return signalPassFailure();
+        rewriter.replaceOp(op, tileAndFuseResult->replacements[op.getResult()]);
+      });
+
+      auto unpackTilingOptions =
+          scf::SCFTilingOptions().setTileSizeComputationFunction(
+              [](OpBuilder &builder, Operation *op) {
+                Location loc = op->getLoc();
+                auto unpackOp = cast<tensor::UnPackOp>(op);
+                int numLoops = unpackOp.getDestRank();
+                auto dimAndTileMapping = unpackOp.getDimAndTileMapping();
+                SmallVector<Value> tileSizes;
+                for (int i = 0; i < numLoops; ++i) {
+                  if (dimAndTileMapping.count(i)) {
```

Comment:
Here I ended-up using tile sizes of 1 too as I did not find a way to generalize unpack this way when outer dims are present.

Reply:
good point, thanks for pointing it out. I also updated the tests which cover the case!